### PR TITLE
Do not use project in do last

### PR DESCRIPTION
**Pull request by reidbaker**
- Use copy task that branches for 8.3 and above to avoid using project.exec in a dolast block. 

part 1/n for flutter/flutter/issues/138523 

This PR does not test the 8.3 branch but instead builds the ability to write a branching gradle test. 
After fixing this project.exec action there I discovered a second that needs to be removed or migrated in order for the build to work. 

Related reading https://docs.gradle.org/8.0/userguide/configuration_cache.html#config_cache:requirements
https://docs.gradle.org/current/javadoc/org/gradle/process/ExecOperations.html (not used but considered) 
https://docs.gradle.org/8.2/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:fileMode

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by reidbaker** [comment]
> Replacement copy works and existing test verifying read only behavior works. I need to add a new test to cover the branching behavior of gradle 8.3 and ensure it does not regress. Thanks to stuart I know I do not need to copy module_test but instead can follow a pattern similar to dev/devicelab/bin/tasks/plugin_test.dart where there a set of tasks with parameters that runs and allows the test to make changes based on the inputs.

Sometime between this comment and now I managed to break the behavior and I cant figure out what I changed.

File: dev/devicelab/bin/tasks/module_test.dart [hunk 1]
```diff
@@ -14,15 +14,37 @@ import 'package:flutter_devicelab/framework/utils.dart';
...
 final String gradlew = Platform.isWindows ? 'gradlew.bat' : 'gradlew';
-final String gradlewExecutable = Platform.isWindows ? '.\\$gradlew' : './$gradlew';
+final String gradlewExecutable =
+    Platform.isWindows ? '.\\$gradlew' : './$gradlew';
 final String fileReadWriteMode = Platform.isWindows ? 'rw-rw-rw-' : 'rw-r--r--';
-final String platformLineSep = Platform.isWindows ? '\r\n': '\n';
+final String platformLineSep = Platform.isWindows ? '\r\n' : '\n';
+
+/// Combines several TaskFunctions with trivial success value into one.
+TaskFunction combine(List<TaskFunction> tasks) {
```
Comment (christopherfujino) [inline dev/devicelab/bin/tasks/module_test.dart:23]: I'm guessing in follow up PRs you will add more tasks here?

File: dev/integration_tests/android_host_app_v2_embedding/gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -2,4 +2,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.3-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-REPLACEME-all.zip
```
Comment (christopherfujino) [inline dev/integration_tests/android_host_app_v2_embedding/gradle/wrapper/gradle-wrapper.properties:5]: was REPLACEME a note to self to change? or are we interpolating this somehow?

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 3]
```diff
@@ -1129,6 +1120,22 @@ class FlutterPlugin implements Plugin<Project> {
...
+                    filePermissions {
+                        user {
+                            read = true
+                            write = true
+                        }
+                    }
+                } else {
+                    // See https://docs.gradle.org/8.2/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:fileMode
+                    // See https://github.com/flutter/flutter/pull/50047
+                    fileMode 0644
```
Comment (christopherfujino) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1137]: this works from windows?

File: dev/integration_tests/android_host_app_v2_embedding/gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (reidbaker) [reply dev/integration_tests/android_host_app_v2_embedding/gradle/wrapper/gradle-wrapper.properties:5]: It is the string we look for when trying to find what to replace.

Comment (christopherfujino) [reply dev/integration_tests/android_host_app_v2_embedding/gradle/wrapper/gradle-wrapper.properties:5]: SGTM

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 3 above)
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1137]: I believe so but dont have a windows machine to check. The test passes on windows, this is the gradle api for setting file mode which works on windows. There is not a windows specific api.

File: dev/devicelab/bin/tasks/module_test.dart (hunk 1 above)
Comment (reidbaker) [reply dev/devicelab/bin/tasks/module_test.dart:23]: Yes once I can make them pass. Pr fatigue is setting in and I want to get this improvement landed to minimize conflicts.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1137]: acknowledged. my understanding is windows' concept of permissions is much less granular than unix and thus there must be some primitive mapping of the octal codes to windows land inside gradle...

...but if the test passes that seems good enough :)

**Review by christopherfujino — APPROVED** [review]
LGTM

File: dev/devicelab/bin/tasks/module_test.dart (hunk 1 above)
Comment (christopherfujino) [reply dev/devicelab/bin/tasks/module_test.dart:23]: SGTM

**Review by bartekpacia — APPROVED** [review]
LGTM, but... wow, these are some wild problems haha

**Comment by asaarnak** [comment]
@reidbaker 

There seems to be an issue when using flutter beta (1a9a60d044)
For example i used gradle version `8.6-rc-2` in `gradle-wrapper.properties`:
```
distributionUrl=https\://services.gradle.org/distributions/gradle-8.6-rc-2-bin.zip
```

And got this error:
```
> Could not create task ':app:copyFlutterAssetsProductionDebug'.
  > For input string: "6-rc-2"
```

**Comment by bartekpacia** [comment]
I think that's because `compareVersionStrings()` doesn't handle `8.6-rc-2`.

**Pull request merged**